packages/tabs/src/Tabs/Tabs.tsx

```diff
@@ -74,62 +67,81 @@ const Tabs = (props: AccessibleTabsProps) => {
 
   const baseFontSize: BaseFontSize = useUpdatedBaseFontSize(baseFontSizeProp);
   const { theme, darkMode } = useDarkMode(darkModeProp);
-  const accessibilityProps = {
-    ['aria-label']: ariaLabel,
-    ['aria-labelledby']: ariaLabelledby,
-  };
-
   const id = useIdAllocator({ prefix: rest.id || 'tabs' });
+
   const { descendants: tabDescendants, dispatch: tabDispatch } =
     useInitDescendants<HTMLDivElement>();
   const { descendants: tabPanelDescendants, dispatch: tabPanelDispatch } =
     useInitDescendants<HTMLDivElement>();
 
+  const isControlled = typeof controlledSelected !== 'undefined';
   const [uncontrolledSelected, setUncontrolledSelected] = useState(0);
   const [selected, setSelected] = [
-    controlledSelected ? controlledSelected : uncontrolledSelected,
-    controlledSelected ? setControlledSelected : setUncontrolledSelected,
+    isControlled ? controlledSelected : uncontrolledSelected,
+    isControlled ? setControlledSelected : setUncontrolledSelected,
   ];
 
-  const childrenArray = useMemo(
-    () => React.Children.toArray(children) as Array<React.ReactElement>,
-    [children],
-  );
+  const accessibilityProps = {
+    ['aria-label']: ariaLabel,
+    ['aria-labelledby']: ariaLabelledby,
+  };
 
-  const handleChange = useCallback(
+  const handleClickTab = useCallback(
     (e: React.SyntheticEvent<Element, MouseEvent>, index: number) => {
       setSelected?.(index);
     },
     [setSelected],
   );
 
-  const getEnabledIndexes: () => [Array<number>, number] = useCallback(() => {
-    const enabledIndexes = childrenArray
-      .filter(child => !child.props.disabled)
-      .map(child => childrenArray.indexOf(child));
-
-    return [enabledIndexes, enabledIndexes.indexOf(selected!)];
-  }, [childrenArray, selected]);
-
-  const handleArrowKeyPress = useCallback(
+  const tabTitleElements = tabDescendants.map(
+    descendant => descendant.element.parentNode as HTMLElement,
+  );
+  const getActiveAndEnabledIndices: () => {
+    activeIndex: number;
+    enabledIndices: Array<number>;
+  } = useCallback(() => {
+    const enabledIndices = tabTitleElements
+      .filter(tabTitleEl => !tabTitleEl.hasAttribute('disabled'))
+      .map(tabTitleEl => tabTitleElements.indexOf(tabTitleEl));
+
+    return {
+      activeIndex: enabledIndices.indexOf(selected),
+      enabledIndices,
+    };
+  }, [selected, tabTitleElements]);
```

Comment:
nit, I think we can make this a bit more "pure-functional". I'd pull out this fn into a separate file, and pass in `tabTitleElements` and `selected` as args

Reply:
moved to a separate file and added tests!